src/mesh.cpp

```diff
@@ -102,6 +121,70 @@ StructuredMesh::bin_label(int bin) const {
   }
 }
 
+//==============================================================================
+// Unstructured Mesh implementation
+//==============================================================================
+
+UnstructuredMesh::UnstructuredMesh(pugi::xml_node node) : Mesh(node) {
+  n_dimension_ = 3;
+
+  // check the mesh type
+  if (check_for_node(node, "type")) {
+    auto temp = get_node_value(node, "type", true, true);
+    if (temp != "unstructured") {
+      fatal_error(fmt::format("Invalid mesh type: {}", temp));
+    }
+  }
+
+  // get the filename of the unstructured mesh to load
+  if (check_for_node(node, "filename")) {
+    filename_ = get_node_value(node, "filename");
+  }
+  else {
+    fatal_error(fmt::format("No filename supplied for unstructured mesh with ID: {}", id_));
+  }
+
+  // check if mesh tally data should be written with
+  // statepoint files
+  if (check_for_node(node, "output")) {
+    output_ = get_node_value_bool(node, "output");
+  }
+
+}
+
+void
+UnstructuredMesh::surface_bins_crossed(const Particle& p,
+                                       std::vector<int>& bins) const {
+  throw std::runtime_error{"Unstructured mesh surface tallies are not implemented."};
+}
+
+std::string
+UnstructuredMesh::bin_label(int bin) const {
+  return fmt::format("Mesh Index ({})", bin);
+};
+
+void
+UnstructuredMesh::to_hdf5(hid_t group) const
+{
+    hid_t mesh_group = create_group(group, fmt::format("mesh {}", id_));
+
+    write_dataset(mesh_group, "type", "unstructured");
+    write_dataset(mesh_group, "filename", filename_);
+    write_dataset(mesh_group, "library", this->library());
+    // write volume of each tet
+    std::vector<double> tet_vols;
+    xt::xtensor<double, 2> centroids({(size_t)this->n_bins(), 3});
+    for (int i = 0; i < this->n_bins(); i++) {
+      tet_vols.emplace_back(this->volume(i));
+      auto c = this->centroid(i);
+      xt::view(centroids, i, xt::all()) = xt::xarray<double>({c.x, c.y, c.z});
+    }
+
+    write_dataset(mesh_group, "volumes", tet_vols);
```

Comment:
I know this was around before, but would you remind me the motivation for writing the volumes and centroid for the UM? It doesn't look like a similar action is performed for the `RectilinearMesh`, for instance. Just wondering if those structured meshes should perhaps be made consistent with the UM case.

Reply:
This provides a way to perform at least a rudimentary visualization of the data that we can support in the Python API without explicit storage of the element vertices, faces, connectivity, etc. 

The second unstructured mesh example demonstrates this (https://docs.openmc.org/en/stable/examples/unstructured-mesh-part-ii.html). It's not ideal, but it's something relatively simple that we can support.

Reply:
Now that I'm thinking about this... given that we have an option for writing a "marked up" version of the mesh. We could write only the mesh in a VTK format and markup the mesh using the python vtk module :thinking:

Reply:
That might be a nice feature - the "rudimentary visualization" you're talking about is the point cloud?

Reply:
Yeah.